### PR TITLE
Use an array for building up mappings in the sourcemap generator.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this
@typescript-bot pack this

File: src/compiler/sourcemap.ts [hunk 1]
```diff
@@ -231,34 +232,37 @@ namespace ts {
...
 
                 if (hasPendingName) {
                     // 5. Relative nameIndex
-                    mappings += base64VLQFormatEncode(pendingNameIndex - lastNameIndex);
+                    mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingNameIndex - lastNameIndex));
                     lastNameIndex = pendingNameIndex;
                 }
             }
 
+            mappings = mappingSegmentsBuffer.join("");
```
Comment (dmichon-msft) [inline src/compiler/sourcemap.ts:263]: In local experimentation, I've found that I can get about a 40% speed up by batching the char code conversion:
```
function base64VLQFormatEncode2(value) {
  if (value < 0) {
      value = ((-value) << 1) + 1;
  }
  else {
      value = value << 1;
  }

  // Encode 5 bits at a time starting from least significant bits
  const result = [];
  do {
      let currentDigit = value & 31; // 11111
      value = value >> 5;
      if (value > 0) {
          // There are still more digits to decode, set the msb (6th bit)
          currentDigit = currentDigit | 32;
      }
      result.push(base64FormatEncode(currentDigit));
  } while (value > 0);

  return result;
}

function buildMappingStr() {
  const charCodes = [];
  // Mock representation of adding a segment, for benchmarking
  for (let i = 0; i < target; i ++) {
    charCodes.push(...base64VLQFormatEncode2(i));
  }
  let str = '';
  const segmentLength = 4096;
  for (let i = 0, len = charCodes.length; i < len; i += segmentLength) {
    const segment = String.fromCharCode.apply(null, charCodes.slice(i, i + segmentLength));
    str += segment;
  }
  return str;
}
```

Comment (dmichon-msft) [reply src/compiler/sourcemap.ts:263]: If you are encoding a very large number of mappings (>10000), it gets faster to use an encoder with its own buffer:
```
class SourceMapEncoder {
  constructor() {
    this._encoded = '';
    this._buffer = new Uint8Array(1024);
    this._bufferPos = 0;
  }

  get encoded() {
    if (this._bufferPos > 0) {
      this._commit(); 
    }
    return this._encoded;
  }

  /**
   * @param {number} code
   */
  appendCharCode(code) {
    this._buffer[this._bufferPos] = code;
    this._bufferPos++;
    
    if (this._bufferPos === this._buffer.length) {
      this._commit();
    }
  }

  /**
   * @param {number} value
   */
  appendBase64VLQ(value) {
    if (value < 0) {
        value = ((-value) << 1) + 1;
    }
    else {
        value = value << 1;
    }
  
    // Encode 5 bits at a time starting from least significant bits
    do {
        let currentDigit = value & 31; // 11111
        value = value >> 5;
        if (value > 0) {
            // There are still more digits to decode, set the msb (6th bit)
            currentDigit = currentDigit | 32;
        }
        this.appendCharCode(formatArr[currentDigit]);
    } while (value > 0);
  }

  reset() {
    this._encoded = '';
    this._bufferPos = 0;
  }

  _commit() {
    this._encoded += String.fromCharCode.apply(null, this._buffer.slice(0, this._bufferPos));
    this._bufferPos = 0;
  }
}
```
Call `.appendBase64VLQ(value)` for each mapping to append, then retrieve `.encoded` at the end.
Still trying to figure out why the overhead is so much higher for small counts

Comment (DanielRosenwasser) [reply src/compiler/sourcemap.ts:263]: That's pretty funny, that's what I've been working on separately over the last hour - check out https://github.com/microsoft/TypeScript/pull/43987

**Comment by sandersn** [comment]
To help with PR housekeeping, I'm going to close this draft PR. It is pretty old now.

**Pull request closed without merging**